File: mcp-server/src/tools/mc-snapshot.ts
// SPDX-License-Identifier: AGPL-3.0-or-later

import type { McpServer } from "@modelcontextprotocol/server";
import * as z from "zod/v4";

import type { PluginClient } from "../plugin-client.js";
import { runTool } from "./helpers.js";

const MAX_VOLUME = 200_000;

const inputSchema = z
    .object({
        action: z
            .enum(["create", "list"])
            .optional()
            .describe('"create" (default) takes a new snapshot; "list" returns every snapshot currently stored.'),
        world: z.string().min(1).optional().describe("World name for a create. Omitted uses the plugin's configured default world."),
        from: z
            .tuple([z.number().int(), z.number().int(), z.number().int()])
            .optional()
            .describe("Inclusive [x, y, z] corner of the region to capture. Required when action is create."),
        to: z
            .tuple([z.number().int(), z.number().int(), z.number().int()])
            .optional()
            .describe("Inclusive [x, y, z] corner opposite `from`. Required when action is create."),
        label: z.string().optional().describe("Optional free-text note stored with the snapshot, e.g. \"before tower\".")
    })
    .refine(v => (v.action ?? "create") !== "create" || (v.from && v.to), {
        message: '"from" and "to" are required when action is "create"'
    });

const DESCRIPTION = `Captures the current block contents of a region so it can be restored later with mc_restore, or lists every snapshot currently stored on the server. This is the manual counterpart to mc_build's own \`snapshot: true\` option - use it to protect an area before an experimental or risky change, independent of any single build call.

WHEN TO USE: before manually editing an area across several separate mc_build calls when you want one restore point covering all of them (mc_build's own snapshot option only covers that single call's bounding box); before letting a user try something destructive; to check what snapshots already exist and their ids/labels/timestamps before deciding whether to restore one, or before running low on the server's snapshot slots (oldest snapshots are evicted automatically once the configured maximum is reached).

WHEN NOT TO USE: to read current block contents (use mc_inspect - snapshots do not display their contents, they only exist to be restored later). If you already passed \`snapshot: true\` to mc_build for a single build call, calling mc_snapshot again for the same region is redundant.

PARAMETERS: \`action\` is "create" (default) or "list". For "create": \`world\` (optional, defaults to the plugin's default world), \`from\`/\`to\` (required, inclusive [x, y, z] corners in any order, X east/Z south/Y up, volume up to 200,000 blocks), and an optional \`label\` string to help identify the snapshot later. "list" takes no other parameters and returns every stored snapshot regardless of world.

SIDE EFFECTS: "create" does not modify the world - it only reads and stores a copy of the region, comparable in cost to mc_inspect over the same volume. It does consume one of the server's limited snapshot slots; the oldest snapshot is silently evicted once the configured maximum count is reached, so do not rely on a snapshot existing indefinitely. "list" has no side effects at all.`;

interface SnapshotResult {
    id: string;
    world: string;
    from: [number, number, number];
    to: [number, number, number];
    volume: number;
    createdAt: string;
    label?: string;
}

interface ListSnapshotsResult {
    snapshots: SnapshotResult[];
}

export function registerMcSnapshot(server: McpServer, client: PluginClient): void {
    server.registerTool(
        "mc_snapshot",
        {
            title: "Snapshot region",
            description: DESCRIPTION,
            inputSchema,
            annotations: { readOnlyHint: false, destructiveHint: false, idempotentHint: false, openWorldHint: false }
        },
        async ({ action, world, from, to, label }) => {
            return runTool(client, "mc_snapshot", async () => {
                if ((action ?? "create") === "list") {
                    const result = (await client.request("list_snapshots", {})) as ListSnapshotsResult;
                    if (result.snapshots.length === 0) {
                        return "No snapshots stored.";
                    }
                    const lines = ["id                          world   from              to                volume  createdAt             label"];
                    for (const s of result.snapshots) {
                        lines.push(
                            `${s.id.padEnd(28)}${s.world.padEnd(8)}${s.from.join(",").padEnd(18)}${s.to.join(",").padEnd(18)}${String(s.volume).padEnd(8)}${s.createdAt.padEnd(22)}${s.label ?? ""}`
                        );
                    }
                    return lines.join("\n");
                }

                if (!from || !to) {
                    throw new Error('"from" and "to" are required when action is "create"');
                }
                const [x1, y1, z1] = [Math.min(from[0], to[0]), Math.min(from[1], to[1]), Math.min(from[2], to[2])];
                const [x2, y2, z2] = [Math.max(from[0], to[0]), Math.max(from[1], to[1]), Math.max(from[2], to[2])];
                const volume = (x2 - x1 + 1) * (y2 - y1 + 1) * (z2 - z1 + 1);
                if (volume > MAX_VOLUME) {
                    throw new Error(
                        `mc_snapshot region volume ${volume} exceeds the 200,000-block limit. Reduce the from/to range or split it into several snapshots.`
                    );
                }

                const snap = (await client.request("snapshot", {
                    world,
                    from: [x1, y1, z1],
                    to: [x2, y2, z2],
                    label
                })) as SnapshotResult;
                return (
                    `Snapshot ${snap.id} created for world "${snap.world}", volume ${snap.volume}, at ${snap.createdAt}.\n` +
                    `Call mc_restore({"id":"${snap.id}"}) to restore this region later.`
                );
            });
        }
    );
}

File: plugin/src/main/java/net/rcwalter/ashlar/tool/mc/McSnapshot.java
// SPDX-License-Identifier: AGPL-3.0-or-later
package net.rcwalter.ashlar.tool.mc;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import net.rcwalter.ashlar.rpc.InvocationContext;
import net.rcwalter.ashlar.rpc.RpcHandler;
import net.rcwalter.ashlar.tool.ArgParse;
import net.rcwalter.ashlar.tool.Tool;
import net.rcwalter.ashlar.tool.ToolArgError;
import net.rcwalter.ashlar.tool.ToolResult;
import net.rcwalter.ashlar.tool.ToolRunner;
import net.rcwalter.ashlar.tool.ToolSpec;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;

import static net.rcwalter.ashlar.tool.mc.JsonUtil.intArray;
import static net.rcwalter.ashlar.tool.mc.JsonUtil.joinArray;
import static net.rcwalter.ashlar.tool.mc.JsonUtil.padEnd;

/** Pure-Java port of {@code mcp-server/src/tools/mc-snapshot.ts}. No text differences from the TS tool. */
public final class McSnapshot implements Tool {

    private static final long MAX_VOLUME = 200_000;
    private static final List<String> ACTIONS = List.of("create", "list");

    private final ToolSpec spec = ToolSpec.load("mc_snapshot");
    private final RpcHandler snapshotHandler;
    private final RpcHandler listSnapshotsHandler;

    public McSnapshot(RpcHandler snapshotHandler, RpcHandler listSnapshotsHandler) {
        this.snapshotHandler = snapshotHandler;
        this.listSnapshotsHandler = listSnapshotsHandler;
    }

    @Override
    public ToolSpec spec() {
        return spec;
    }

    /** Client-side pre-check mirroring mc-snapshot.ts's {@code MAX_VOLUME} guard, exposed for unit testing. */
    static void checkVolume(long volume) {
        if (volume > MAX_VOLUME) {
            throw new ToolArgError("mc_snapshot region volume " + volume
                    + " exceeds the 200,000-block limit. Reduce the from/to range or split it into several snapshots.");
        }
    }

    record Args(String action, String world, int[] from, int[] to, String label) {
        static Args parse(JsonObject o) {
            String action = ArgParse.optEnum(o, "action", ACTIONS, "create");
            String world = ArgParse.optString(o, "world");
            int[] from = ArgParse.has(o, "from") ? ArgParse.requireCoords3(o, "from") : null;
            int[] to = ArgParse.has(o, "to") ? ArgParse.requireCoords3(o, "to") : null;
            String label = ArgParse.optString(o, "label");
            if (action.equals("create") && (from == null || to == null)) {
                throw new ToolArgError("\"from\" and \"to\" are required when action is \"create\"");
            }
            return new Args(action, world, from, to, label);
        }
    }

    @Override
    public CompletableFuture<ToolResult> call(InvocationContext ctx, JsonObject args) {
        return ToolRunner.runText("mc_snapshot", () -> {
            Args a = Args.parse(args);

            if (a.action().equals("list")) {
                return listSnapshotsHandler.handle(ctx, new JsonObject()).thenApply(el -> {
                    JsonArray snaps = el.getAsJsonObject().getAsJsonArray("snapshots");
                    if (snaps.isEmpty()) {
                        return "No snapshots stored.";
                    }
                    List<String> lines = new ArrayList<>();
                    lines.add("id                          world   from              to                volume  createdAt             label");
                    for (JsonElement se : snaps) {
                        JsonObject s = se.getAsJsonObject();
                        lines.add(padEnd(s.get("id").getAsString(), 28)
                                + padEnd(s.get("world").getAsString(), 8)
                                + padEnd(joinArray(s.getAsJsonArray("from")), 18)
                                + padEnd(joinArray(s.getAsJsonArray("to")), 18)
                                + padEnd(String.valueOf(s.get("volume").getAsLong()), 8)
                                + padEnd(s.get("createdAt").getAsString(), 22)
                                + (s.has("label") && !s.get("label").isJsonNull() ? s.get("label").getAsString() : ""));
                    }
                    return String.join("\n", lines);
                });
            }

            if (a.from() == null || a.to() == null) {
                throw new ToolArgError("\"from\" and \"to\" are required when action is \"create\"");
            }
            int x1 = Math.min(a.from()[0], a.to()[0]), y1 = Math.min(a.from()[1], a.to()[1]), z1 = Math.min(a.from()[2], a.to()[2]);
            int x2 = Math.max(a.from()[0], a.to()[0]), y2 = Math.max(a.from()[1], a.to()[1]), z2 = Math.max(a.from()[2], a.to()[2]);
            long volume = (long) (x2 - x1 + 1) * (y2 - y1 + 1) * (z2 - z1 + 1);
            checkVolume(volume);

            JsonObject params = new JsonObject();
            if (a.world() != null) {
                params.addProperty("world", a.world());
            }
            params.add("from", intArray(x1, y1, z1));
            params.add("to", intArray(x2, y2, z2));
            if (a.label() != null) {
                params.addProperty("label", a.label());
            }

            return snapshotHandler.handle(ctx, params).thenApply(el -> {
                JsonObject snap = el.getAsJsonObject();
                return "Snapshot " + snap.get("id").getAsString() + " created for world \"" + snap.get("world").getAsString()
                        + "\", volume " + snap.get("volume").getAsLong() + ", at " + snap.get("createdAt").getAsString() + ".\n"
                        + "Call mc_restore({\"id\":\"" + snap.get("id").getAsString() + "\"}) to restore this region later.";
            });
        });
    }
}
